File: scripts/youtube_api_upload.py
#!/usr/bin/env python3
import argparse
import os
import sys

from google.auth.transport.requests import Request
from google.auth.exceptions import RefreshError
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
from googleapiclient.http import MediaFileUpload

SCOPES = [
    "https://www.googleapis.com/auth/youtube.upload",
    "https://www.googleapis.com/auth/youtube.readonly",
    # Required for playlistItems.insert (add uploaded video to a playlist).
    "https://www.googleapis.com/auth/youtube.force-ssl",
]


def get_credentials(client_secrets: str, token_file: str) -> Credentials:
    creds = None
    if os.path.exists(token_file):
        try:
            creds = Credentials.from_authorized_user_file(token_file, SCOPES)
            if not creds.has_scopes(SCOPES):
                creds = None
        except Exception:
            creds = None

    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            try:
                creds.refresh(Request())
            except RefreshError as exc:
                # invalid_grant means refresh token was revoked/expired; re-auth interactively.
                if "invalid_grant" not in str(exc):
                    raise
                creds = None
        if not creds:
            flow = InstalledAppFlow.from_client_secrets_file(client_secrets, SCOPES)
            creds = flow.run_local_server(port=0)
        os.makedirs(os.path.dirname(token_file), exist_ok=True)
        with open(token_file, "w", encoding="utf-8") as f:
            f.write(creds.to_json())

    return creds


def add_video_to_playlist(youtube, playlist_id: str, video_id: str) -> None:
    youtube.playlistItems().insert(
        part="snippet",
        body={
            "snippet": {
                "playlistId": playlist_id,
                "resourceId": {"kind": "youtube#video", "videoId": video_id},
            }
        },
    ).execute()


def main() -> int:
    parser = argparse.ArgumentParser(description="Upload one video to YouTube.")
    parser.add_argument("--file", default="")
    parser.add_argument("--title", default="")
    parser.add_argument("--description", default="")
    parser.add_argument("--category-id", default="2")
    parser.add_argument("--privacy", default="unlisted", choices=["public", "unlisted", "private"])
    parser.add_argument("--client-secrets", required=True)
    parser.add_argument("--token-file", required=True)
    parser.add_argument("--target-channel-id", default="")
    parser.add_argument(
        "--playlist-id",
        default="",
        help="If set, add the new video to this playlist (ID from studio URL, e.g. PL…).",
    )
    parser.add_argument("--check-channel-only", action="store_true")
    args = parser.parse_args()

    if not os.path.exists(args.client_secrets):
        print(f"Client secrets not found: {args.client_secrets}", file=sys.stderr)
        return 1

    try:
        creds = get_credentials(args.client_secrets, args.token_file)
        youtube = build("youtube", "v3", credentials=creds)
        channels = youtube.channels().list(part="id,snippet", mine=True).execute()
        items = channels.get("items", [])
        channel_ids = [item.get("id", "") for item in items]

        if not channel_ids:
            print("No accessible YouTube channel found for authenticated account.", file=sys.stderr)
            return 1

        if args.target_channel_id and args.target_channel_id not in channel_ids:
            print(
                f"Authenticated channel does not match YT_TARGET_CHANNEL_ID={args.target_channel_id}. "
                f"Available: {', '.join(channel_ids)}",
                file=sys.stderr,
            )
            return 1

        active = items[0]
        print(f"Authenticated channel: {active.get('snippet', {}).get('title', '')} ({active.get('id', '')})")

        if args.check_channel_only:
            return 0

        if not args.file:
            print("--file is required unless --check-channel-only is used.", file=sys.stderr)
            return 1
        if not args.title:
            print("--title is required unless --check-channel-only is used.", file=sys.stderr)
            return 1
        if not os.path.exists(args.file):
            print(f"File not found: {args.file}", file=sys.stderr)
            return 1

        body = {
            "snippet": {
                "title": args.title,
                "description": args.description,
                "categoryId": args.category_id,
            },
            "status": {"privacyStatus": args.privacy},
        }
        media = MediaFileUpload(args.file, resumable=True)
        request = youtube.videos().insert(part="snippet,status", body=body, media_body=media)
        response = request.execute()
        video_id = response.get("id", "")
        print(f"Uploaded video id: {video_id}")

        playlist_id = (args.playlist_id or "").strip()
        if playlist_id and video_id:
            try:
                add_video_to_playlist(youtube, playlist_id, video_id)
                print(f"Added to playlist: {playlist_id}")
            except HttpError as exc:
                print(
                    f"Upload succeeded but playlist add failed (video is live): {exc}",
                    file=sys.stderr,
                )
                print(
                    "Fix YT_PLAYLIST_ID / permissions or add the video manually in YouTube Studio.",
                    file=sys.stderr,
                )
        return 0
    except HttpError as exc:
        print(f"YouTube API error: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/youtube_api_upload.js
#!/usr/bin/env node
const fs = require("fs");
const path = require("path");
const readline = require("readline");
const { google } = require("googleapis");
const { spawnSync } = require("child_process");

const SCOPES = [
  "https://www.googleapis.com/auth/youtube.upload",
  "https://www.googleapis.com/auth/youtube.readonly",
  "https://www.googleapis.com/auth/youtube.force-ssl",
];

function parseArgs(argv) {
  const out = {
    file: "",
    title: "",
    description: "",
    categoryId: "2",
    privacy: "unlisted",
    clientSecrets: "",
    tokenFile: "",
    targetChannelId: "",
    playlistId: "",
    checkChannelOnly: false,
  };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    const next = argv[i + 1];
    if (arg === "--file") out.file = next || "";
    else if (arg === "--title") out.title = next || "";
    else if (arg === "--description") out.description = next || "";
    else if (arg === "--category-id") out.categoryId = next || "2";
    else if (arg === "--privacy") out.privacy = next || "unlisted";
    else if (arg === "--client-secrets") out.clientSecrets = next || "";
    else if (arg === "--token-file") out.tokenFile = next || "";
    else if (arg === "--target-channel-id") out.targetChannelId = next || "";
    else if (arg === "--playlist-id") out.playlistId = next || "";
    else if (arg === "--check-channel-only") out.checkChannelOnly = true;
    if (arg.startsWith("--") && arg !== "--check-channel-only") i += 1;
  }
  return out;
}

function prompt(question) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise((resolve) => {
    rl.question(question, (answer) => {
      rl.close();
      resolve(answer.trim());
    });
  });
}

function openUrl(url) {
  try {
    if (process.platform === "darwin") spawnSync("open", [url], { stdio: "ignore" });
    else if (process.platform === "win32") spawnSync("cmd", ["/c", "start", "", url], { stdio: "ignore" });
    else spawnSync("xdg-open", [url], { stdio: "ignore" });
  } catch (e) {
    // ignore open failures
  }
}

function loadClientSecrets(file) {
  const expanded = path.resolve(file.replace(/^~(?=$|\/|\\)/, require("os").homedir()));
  if (!fs.existsSync(expanded)) throw new Error(`Client secrets not found: ${expanded}`);
  const json = JSON.parse(fs.readFileSync(expanded, "utf8"));
  const cfg = json.installed || json.web;
  if (!cfg || !cfg.client_id || !cfg.client_secret) {
    throw new Error("Invalid client secrets JSON: expected installed/client_id/client_secret");
  }
  return { cfg, resolvedPath: expanded };
}

async function getAuthorizedClient(clientSecretsPath, tokenFilePath) {
  const { cfg } = loadClientSecrets(clientSecretsPath);
  const tokenPath = path.resolve(tokenFilePath.replace(/^~(?=$|\/|\\)/, require("os").homedir()));
  const oauth2Client = new google.auth.OAuth2(
    cfg.client_id,
    cfg.client_secret,
    (cfg.redirect_uris && cfg.redirect_uris[0]) || "http://localhost"
  );

  let existing = null;
  if (fs.existsSync(tokenPath)) {
    try {
      existing = JSON.parse(fs.readFileSync(tokenPath, "utf8"));
      oauth2Client.setCredentials(existing);
    } catch (e) {
      existing = null;
    }
  }

  try {
    await oauth2Client.getAccessToken();
    return oauth2Client;
  } catch (e) {
    // Need interactive auth
  }

  const authUrl = oauth2Client.generateAuthUrl({
    access_type: "offline",
    scope: SCOPES,
    prompt: "consent",
  });
  console.error("Authorize this app by visiting this URL:");
  console.error(authUrl);
  const openNow = await prompt("Open URL in browser now? [Y/n]: ");
  if (openNow === "" || openNow.toLowerCase() === "y" || openNow.toLowerCase() === "yes") {
    openUrl(authUrl);
  }
  const code = await prompt("Paste authorization code here: ");
  const tokenResult = await oauth2Client.getToken(code);
  const tokens = tokenResult.tokens || {};
  if (!tokens.refresh_token && existing && existing.refresh_token) {
    tokens.refresh_token = existing.refresh_token;
  }
  oauth2Client.setCredentials(tokens);
  fs.mkdirSync(path.dirname(tokenPath), { recursive: true });
  fs.writeFileSync(tokenPath, `${JSON.stringify(tokens, null, 2)}\n`, "utf8");
  return oauth2Client;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));
  if (!args.clientSecrets) throw new Error("--client-secrets is required");
  if (!args.tokenFile) throw new Error("--token-file is required");
  if (!["public", "unlisted", "private"].includes(args.privacy)) {
    throw new Error(`Invalid --privacy: ${args.privacy}`);
  }

  const auth = await getAuthorizedClient(args.clientSecrets, args.tokenFile);
  const youtube = google.youtube({ version: "v3", auth });

  const channels = await youtube.channels.list({ part: ["id", "snippet"], mine: true });
  const items = channels.data.items || [];
  const channelIds = items.map((item) => item.id || "").filter(Boolean);
  if (channelIds.length === 0) {
    throw new Error("No accessible YouTube channel found for authenticated account.");
  }
  if (args.targetChannelId && !channelIds.includes(args.targetChannelId)) {
    throw new Error(
      `Authenticated channel does not match YT_TARGET_CHANNEL_ID=${args.targetChannelId}. Available: ${channelIds.join(", ")}`
    );
  }
  const active = items[0];
  console.log(`Authenticated channel: ${active.snippet?.title || ""} (${active.id || ""})`);

  if (args.checkChannelOnly) return;
  if (!args.file) throw new Error("--file is required unless --check-channel-only is used.");
  if (!args.title) throw new Error("--title is required unless --check-channel-only is used.");
  if (!fs.existsSync(args.file)) throw new Error(`File not found: ${args.file}`);

  const upload = await youtube.videos.insert({
    part: ["snippet", "status"],
    requestBody: {
      snippet: {
        title: args.title,
        description: args.description || "",
        categoryId: args.categoryId || "2",
      },
      status: { privacyStatus: args.privacy },
    },
    media: {
      body: fs.createReadStream(args.file),
    },
  });
  const videoId = upload.data.id || "";
  console.log(`Uploaded video id: ${videoId}`);

  if (args.playlistId && videoId) {
    try {
      await youtube.playlistItems.insert({
        part: ["snippet"],
        requestBody: {
          snippet: {
            playlistId: args.playlistId,
            resourceId: { kind: "youtube#video", videoId },
          },
        },
      });
      console.log(`Added to playlist: ${args.playlistId}`);
    } catch (e) {
      console.error(`Upload succeeded but playlist add failed (video is live): ${e.message || e}`);
      console.error("Fix YT_PLAYLIST_ID / permissions or add the video manually in YouTube Studio.");
    }
  }
}

main().catch((err) => {
  console.error(`YouTube API error: ${err.message || err}`);
  process.exit(1);
});
